feat(KFLUXUI-231): [Release Details] add `YAML` tab

rrosatti opened the pull request:
## Fixes 

Fixes https://issues.redhat.com/browse/KFLUXUI-231


## Description

This PR adds a "YAML" tab to the Release Details page. The tab allows the users to view the full release YAML in a read-only editor.

The implementation uses [PatternFly's @patternfly/react-code-editor](https://www.npmjs.com/package/@patternfly/react-code-editor)(which uses Monaco Editor behind the scenes). This setup provides syntax highlighting, validation, and a read-only experience for YAML content.

Additional libraries used:

* `yaml` – for parsing and formatting the YAML data
* `monaco-editor` – core editor
* `monaco-yaml` – YAML language support and schema validation
* `monaco-editor-webpack-plugin` – proper bundling and worker configuration
* `openapi-types` – typing support for OpenAPI schema parsing


## Type of change
<!-- Please delete options that are not relevant. -->

- [x] Feature
- [ ] Bugfix
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Screen shots / Gifs for design review 

https://github.com/user-attachments/assets/dbdd1122-02d5-4ebd-8731-a451bbd5ae76

## How to test or reproduce?
<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

1. Go to "Releases" page
2. Select a release from the list
3. Click on `YAML` tab
4. You should see a read-only Monaco Editor showing the YAML content of the selected release
5. ☕ 

<!-- **Test Configuration(s)**: -->


## Browser conformance: 
<!-- To mark tested browsers, use [x] -->
- [x] Chrome
- [ ] Firefox
- [x] Safari
- [ ] Edge


<!-- ## Checklist: -->

<!-- 
- [ ] Code follows the style guidelines
- [ ] Self-reviewed the code
- [ ] Added comments in hard-to-understand areas
- [ ] Made corresponding changes to the documentation
- [ ] Changes generate no new warnings
- [ ] Added tests that prove this fix is effective or that the feature works
- [ ] New and existing unit tests pass locally with new changes
- [ ] Any dependent changes have been merged and published in downstream modules 
-->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a YAML tab to Release Details with a read-only, schema-aware YAML editor (syntax highlighting, validation, autocompletion), loading/error states, and keyboard-shortcuts UI.
  * Added a reusable code/YAML editor component for viewing resource YAML.

* **Tests**
  * Added extensive unit tests covering the YAML tab, editor, schema integration, swagger fetching, and shortcut UI.

* **Chores**
  * Integrated editor packages and build config to support the Monaco-based YAML editor.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rrosatti committed:
chore: add editor and YAML-related dependencies

Added the following libraries:
- `@patternfly/react-code-editor`: wrapper component around Monaco Editor for
integration with PatternFly.
- `yaml`: for parsing and stringifying YAML content.
- `monaco-editor`: core Monaco Editor package
- `monaco-editor-webpack-plugin`: enables custom Monaco configuration and
language support via Webpack
- `monaco-yaml`: provides YAML language support for Monaco Editor
- `openapi-types`: type definitions for OpenAPI schema support.

rrosatti committed:
feat(Releases/ReleaseYamlTab): add tab to view release YAML

Added a new "Release YAML" tab to display the YAML representation of the
respective release.

- uses the `yaml` library to stringify the release object
- displays the YAML content in a "Monaco Editor" component with read-only mode.

rrosatti committed:
feat(Releases/__tests__): add tests for ReleaseYamlTab component

Add a few tests to cover some rendering scenarios of the 'ReleaseYamlTab'
component.

rrosatti commented on jest.config.js:
```diff
@@ -22,7 +22,7 @@ export default {
   transformIgnorePatterns: [
-    '/node_modules/(?!@patternfly|uuid|lodash-es|@popperjs|i18next|d3|d3-array|delaunator|robust-predicates|internmap|react-dnd|react-dnd-html5-backend|dnd-core|@react-dnd)',
+    '/node_modules/(?!@patternfly|uuid|lodash-es|@popperjs|i18next|d3|d3-array|delaunator|robust-predicates|internmap|react-dnd|react-dnd-html5-backend|dnd-core|@react-dnd|yaml)',
   ],
```
this is to fix the following error when running `yarn test`:

```bash
konflux-ui/node_modules/yaml/browser/index.js:3
import * as YAML from './dist/index.js'
^^^^^^

SyntaxError: Cannot use import statement outside a module
```

rrosatti committed:
feat(shared/components/code-editor): add support for YAML file display

Introduce new components based on PatternFly's CodeEditor to render
code files. Currently, only ".yaml" files are supported via the
YAMLCodeEditor component. Includes reusable hooks, types, and shortcuts
to support editor functionality.

rrosatti committed:
test(shared/components/code-editor): add tests for YAMLCodeEditor and CodeEditor

Add a few tests to cover the usability of both 'YAMLCodeEditor' and
'CodeEditor' components.

rrosatti committed:
chore(webpack): configure monaco-editor to load locally with custom workers

Set up Monaco Editor to load locally instead of via CDN by configuring
"@monaco-editor/react" with the "monaco-editor" package and
"monaco-editor-webpack-plugin".
This includes adding support for custom languages (e.g., YAML) with custom
workers, and adjusting style-loader rules to handle Monaco’s CSS separately.

rrosatti commented on package.json:
```diff
@@ -87,6 +89,10 @@
     "jest-environment-jsdom": "^29.7.0",
     "lint-staged": "^15.2.7",
     "mini-css-extract-plugin": "^2.9.0",
+    "monaco-editor": "^0.51.0",
```
I mentioned in the PR description, but this is a quick note about why we're adding these libs:

* `yaml` – for parsing and formatting the YAML data
* `monaco-editor` – core editor
* `monaco-yaml` – YAML language support and schema validation
* `monaco-editor-webpack-plugin` – proper bundling and worker configuration
* `openapi-types` – typing support for OpenAPI schema parsing

rrosatti commented on src/shared/components/code-editor/CodeEditor.tsx:
```diff
@@ -0,0 +1,46 @@
+import * as monacoConfig from 'monaco-editor/esm/vs/editor/editor.api';
+import { useShortcutPopover } from './hooks/useShortcutPopover';
+
+loader.config({ monaco: monacoConfig });
```
this is to avoid using monaco from CDN - you can check it out in the lib's main README -> https://www.npmjs.com/package/@patternfly/react-code-editor

![Screenshot 2025-04-23 at 16 18 53](https://github.com/user-attachments/assets/c23bfea1-00b5-4d0d-85ec-e47e4d2ce9e4)

rrosatti commented on src/shared/components/code-editor/hooks/useSwaggerDefinitions.ts:
```diff
@@ -0,0 +1,31 @@
+
+const fetchSwaggerDefinitions = async (): Promise<OpenAPISchemaDefinitions> => {
+  try {
+    const response = await commonFetch('/openapi/v2');
```
we're fetching the OpenAPI v2 definitions to configure the Monaco editor. This allows the YAML editor to show property descriptions when hovering over it:

![Screenshot 2025-04-23 at 16 23 40](https://github.com/user-attachments/assets/6bb418bf-a524-4fee-86cb-42e098162923)

rrosatti commented on src/shared/components/code-editor/openapi-to-json-schema.ts:
```diff
@@ -0,0 +1,124 @@
+/* eslint-disable @typescript-eslint/no-unsafe-argument */
+/* eslint-disable no-prototype-builtins */
+
+// this file was copied from openshift/console repository - it has some util functions to convert from
```
as I mentioned here, this file was copied from the OpenShift Console repo - it works the same as in our case, so I opted not to rewrite or modify it for now. But if you guys think that's not a good approach, I'm totally open to suggestions :)

sahil143 replied:
Which package from openshift/console has this configuration?

rrosatti commented on webpack.dev.config.js:
```diff
@@ -93,7 +100,29 @@ export default merge(commonConfig, {
+  plugins: [
+    new ReactRefreshWebpackPlugin(),
+    new ForkTsCheckerWebpackPlugin({ devServer: false }),
+    new MonacoWebpackPlugin({
```
this webpack plugin config is also based on what the OpenShift Console team is using on their side

sahil143 replied:
I think this should be part of common webpack, this plugin would be required for prod also.

sahil143 commented on src/components/Releases/ReleaseYamlTab.tsx:
```diff
@@ -0,0 +1,31 @@
+  }
+
+  return (
+    <div>
+      <Flex className="pf-v5-u-py-lg">
+        <YAMLCodeEditor code={release} data-test="monaco-editor" />
+      </Flex>
+    </div>
```
Do we need this wrapper div?

sahil143 commented on src/shared/components/code-editor/hooks/useSwaggerDefinitions.ts:
```diff
@@ -0,0 +1,31 @@
+
+export const useSwaggerDefinitions = () => {
+  return useQuery(
+    queryOptions({
+      queryKey: [SWAGGER_QUERY_KEY],
+      queryFn: () => fetchSwaggerDefinitions(),
+    }),
+  );
```
pass a stale time of Infinity. so we dont need to fetch this again once fetched

sahil143 commented on src/shared/components/code-editor/CodeEditor.tsx:
```diff
@@ -0,0 +1,46 @@
+        }}
+        onEditorDidMount={onEditorDidMount}
+        isDarkTheme
+        height="500px"
```
We have minHeight applied on the wrapper div, Do we need this height also?

sahil143 commented on src/shared/components/code-editor/CodeEditorShortcut.tsx:
```diff
@@ -0,0 +1,39 @@
+
+const CodeEditorShortcut: React.FC<CodeEditorShortcutProps> = ({ children, hover, keyName }) => {
+  return (
+    <tr>
```
Use Grid here instead of table?

sahil143 commented on src/routes/page-routes/release.tsx:
```diff
@@ -18,6 +19,7 @@ const releaseRoutes = [
         index: true,
         element: <ReleaseOverviewTab />,
       },
+      { path: 'yaml', element: <ReleaseYamlTab /> },
```
Lazy load this component

rrosatti committed:
feat(Releases): add "YAML" tab to release details page

This patch has the necessary changes to display the "YAML" tab in the "Release
Details" page.

testcara commented on src/shared/components/code-editor/__tests__/YAMLCodeEditor.spec.tsx:
```diff
@@ -0,0 +1,107 @@
+describe('YAMLCodeEditor', () => {
+  beforeEach(() => {
+    (useSwaggerDefinitions as jest.Mock).mockReturnValue({
+      data: {
```
it seems the data is also enjoyed by ReleaseYamlTab.spec.tsx.  How about we create data file to ensure they are able to share the vars ?

rrosatti replied:
nice suggestion, updated!

testcara reviewed (dismissed):
just one tiny suggestion for testing data. no matter it is changed or not, the patch is
LGTM to me.

testcara reviewed (dismissed):
lgtm

sahil143 commented on src/routes/page-routes/release.tsx:
```diff
@@ -18,6 +18,13 @@ const releaseRoutes = [
+      {
+        path: 'yaml',
+        async lazy() {
+          const { ReleaseYamlTab } = await import('../../components/Releases');
```
```suggestion
          const { ReleaseYamlTab } = await import('../../components/Releases/ReleaseYamlTab');
```
would be better?

sahil143 commented on src/components/Releases/index.ts:
```diff
@@ -30,5 +30,6 @@ export const releaseDetailsViewLoader = createLoaderWithAccessCheck(
 );
 
 export { default as ReleaseOverviewTab } from './ReleaseOverviewTab';
+export { default as ReleaseYamlTab } from './ReleaseYamlTab';
```
Do we need this if you import the component directly from ReleaseYamlTab?

sahil143 reviewed (changes requested):
Can you check the increase in bundle size after adding the monaco dependency and if lazy loading a component is working as expected?

StanislavJochman reviewed (dismissed):
LGTM

janaki29 reviewed (dismissed):
LGTM

rrosatti commented:
hey @sahil143 - do you have any suggestions about what should we do with this PR? :) I believe it's pretty much ready, the only thing we were discussing before was regarding the bundle size. I'm not sure how much we could still improve it, tho.

rakshett reviewed (dismissed):
LGTM

testcara reviewed (dismissed):
LGTM

StanislavJochman reviewed (dismissed):
LGTM

StanislavJochman reviewed (dismissed):
LGTM

testcara commented on src/shared/components/code-editor/CodeEditorShortcut.scss:
```diff
@@ -0,0 +1,23 @@
+
+  &__kbd {
+    border: var(--pf-v5-global--BorderWidth--sm) solid var(--pf-v5-global--BorderColor--100);
+    border-radius: 3px;
```
border-radius: var(--pf-v5-global--BorderRadius--sm);
font-size: var(--pf-v5-global--FontSize--xs);
padding: var(--pf-v5-global--spacer--xs) var(--pf-v5-global--spacer--xxs); ?

testcara reviewed (dismissed):
LGTM, works well locally.
But i am not sure whether 'CodeMirror' is more suitable for the situation.
We just need to support simple config-map edition. Manoco + code editor is a bit too heavy.

rrosatti commented on webpack.config.js:
```diff
@@ -53,10 +54,30 @@ export default {
+        test: /\.css$/,
+        include: path.resolve(__dirname, 'node_modules/monaco-editor'),
+        use: ['style-loader', 'css-loader'],
+      },
```
I tried to remove it, but I'm getting the following error:

![Screenshot 2025-12-18 at 12 18 19](https://github.com/user-attachments/assets/c2efa1b1-7f95-482e-99ca-56ad232c54f2)

sahil143 replied:
you need to include the monaco editor css file in the `\.s?[ac]ss$/i,` config by default it will exclude node module if it's not imported anywhere in the code

JoaoPedroPP reviewed (dismissed):
LGTM

UI is working as expected

StanislavJochman reviewed (dismissed):
LGTM

testcara reviewed (dismissed):
LGTM, UI works as expected.

sahil143 commented on src/shared/components/code-editor/hooks/useSwaggerDefinitions.ts:
```diff
@@ -0,0 +1,30 @@
+    return data.definitions as OpenAPISchemaDefinitions;
+  } catch (e) {
+    logger.error('Could not get OpenAPI definitions', e instanceof Error ? e : undefined);
+    return null;
```
**Critical:** `queryFn` catches all errors and returns `null` — React Query will set `status: 'success'` and `error: null`. This means the `if (error)` branch in `YAMLCodeEditor.tsx:37` is **unreachable in production**.

The error tests in `YAMLCodeEditor.spec.tsx` only pass because they manually mock `error` to a truthy value, but the real hook never produces that state.

Either:
- Re-throw after logging so React Query tracks the error and the error UI works: `logger.error(...); throw e;`
- Or, if the intent is graceful degradation (show editor without schema on failure), remove the error handling from `YAMLCodeEditor` and document that schema loading is best-effort.

The current code pretends to handle errors but doesn't.

sahil143 commented on src/shared/components/code-editor/monaco.ts:
```diff
@@ -0,0 +1,47 @@
+) => {
+  if (!monacoInstance) {
+    // eslint-disable-next-line no-console
+    console.warn('registerYAMLinMonaco: monacoInstance is undefined');
```
**Bug:** `console.warn` in production code — the project enforces `no console.*` via ESLint. The same PR correctly uses `logger.error` in `useSwaggerDefinitions.ts`. This should be `logger.warn(...)` from `~/monitoring/logger`.

sahil143 commented on src/components/Releases/__tests__/ReleaseYamlTab.spec.tsx:
```diff
@@ -0,0 +1,81 @@
+import React from 'react';
+import { screen } from '@testing-library/react';
+import { stringify as yamlStringify } from 'yaml';
+import { useRelease } from '../../../hooks/useReleases';
```
**Suggestion:** Relative imports — every other test in `src/components/Releases/__tests__/` uses absolute `~/` imports. These should be:
```ts
import { useRelease } from '~/hooks/useReleases';
import { mockedSwaggerDefinitions } from '~/shared/components/code-editor/__data__/mock-data';
import { useSwaggerDefinitions } from '~/shared/components/code-editor/hooks/useSwaggerDefinitions';
import { renderWithQueryClient } from '~/unit-test-utils';
import { mockUseNamespaceHook } from '~/unit-test-utils/mock-namespace';
```

The `jest.mock()` calls below should also use `~/` paths.

sahil143 commented on src/shared/components/code-editor/YAMLCodeEditor.tsx:
```diff
@@ -0,0 +1,51 @@
+    }
+  }, [monacoInstance, swaggerDefinitions]);
+
+  if (isLoading) {
```
**Suggestion:** The entire editor shows a spinner while the OpenAPI spec loads. The Kubernetes OpenAPI v2 spec is typically 10-20MB of JSON, so this could block the tab for several seconds.

Consider rendering the editor immediately with syntax highlighting and loading schema features (hover, validation) in the background. The `registerYAMLinMonaco` call already runs via `useEffect` after mount — the editor could simply render without schema support initially, and gain it once the fetch completes.

sahil143 commented on src/shared/components/code-editor/monaco.ts:
```diff
@@ -0,0 +1,47 @@
+   * We check that > 1 YAML language exists because one is the default and
+   * the other is the language server that we register.
+   */
+  if ((monacoInstance.languages?.getLanguages() ?? []).filter((x) => x.id === 'yaml').length <= 1) {
```
**Suggestion:** Counting YAML language entries to detect prior registration depends on `configureMonacoYaml` adding a second entry — an implementation detail of `monaco-yaml` that could change.

A module-level boolean flag would be simpler and more reliable:
```ts
let yamlRegistered = false;

export const registerYAMLinMonaco = (...) => {
  if (!monacoInstance) { ... }
  if (yamlRegistered) return;
  yamlRegistered = true;
  // ...
};
```

sahil143 commented on src/components/Releases/ReleaseYamlTab.tsx:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  return (
+    <Flex className="pf-v5-u-py-lg">
```
**Suggestion:** Consider using PatternFly v5 design tokens via a co-located SCSS class instead of the utility class string:
```scss
.release-yaml-tab {
  padding-block: var(--pf-v5-global--spacer--lg);
}
```
Or at minimum use the style prop with the design token:
```tsx
<Flex style={{ paddingBlock: 'var(--pf-v5-global--spacer--lg)' }}>
```
The review guidelines flag both utility class strings and inline `style={{}}` as patterns to avoid in favor of co-located SCSS.

sahil143 commented on webpack.config.js:
```diff
@@ -58,6 +59,21 @@ export default {
+    new MonacoWebpackPlugin({
+      languages: ['yaml'], // add only 'yaml' language, since that's the only one we're going to support for now
+      features: ['find', 'quickCommand', 'hover'], // only basic features we might need for now
+      globalAPI: true,
```
**Suggestion:** `globalAPI: true` exposes Monaco on `window.monaco`, which is the reason the duplicate-registration guard in `monaco.ts` is needed. Worth investigating whether `globalAPI: false` works — `loader.config({ monaco: monacoConfig })` in `CodeEditor.tsx` should be sufficient. If it works, the registration guard logic in `monaco.ts` becomes unnecessary and you avoid global scope pollution.

sahil143 reviewed (changes requested):
Thanks for this feature — the YAML tab with schema-aware hover is a solid UX addition. I have a few findings, one critical bug in the error handling and several suggestions.

rrosatti committed:
fixup! chore(webpack): configure monaco-editor to load locally with custom workers

rrosatti committed:
fixup! feat(shared/components/code-editor): add support for YAML file display

rrosatti committed:
fixup! test(shared/components/code-editor): add tests for YAMLCodeEditor and CodeEditor

rrosatti committed:
fixup! feat(Releases/ReleaseYamlTab): add tab to view release YAML

rrosatti committed:
fixup! feat(Releases/__tests__): add tests for ReleaseYamlTab component

rrosatti committed:
fixup! test(shared/components/code-editor): add tests for YAMLCodeEditor and CodeEditor

rrosatti committed:
fixup! test(shared/components/code-editor): add tests for YAMLCodeEditor and CodeEditor

StanislavJochman reviewed (approved):
LGTM
